src/tools/id-mapping/components/IDMappingForm.tsx

```diff
@@ -0,0 +1,455 @@
+import {
+  useRef,
+  FormEvent,
+  useState,
+  useMemo,
+  useEffect,
+  useCallback,
+} from 'react';
+import { useDispatch } from 'react-redux';
+import { v1 } from 'uuid';
+import {
+  PageIntro,
+  Message,
+  TreeSelect,
+  SpinnerIcon,
+  Loader,
+} from 'franklin-sites';
+import { sleep } from 'timing-functions';
+import { useHistory } from 'react-router-dom';
+
+import useDataApi from '../../../shared/hooks/useDataApi';
+import useTextFileInput from '../../../shared/hooks/useTextFileInput';
+import useReducedMotion from '../../../shared/hooks/useReducedMotion';
+
+import AutocompleteWrapper from '../../../query-builder/components/AutocompleteWrapper';
+
+import { createJob } from '../../state/toolsActions';
+import { parseIDs, joinIDs, getTreeData } from '../utils';
+import { truncateTaxonLabel } from '../../utils';
+
+import infoMappings from '../../../shared/config/InfoMappings';
+import apiUrls from '../../../shared/config/apiUrls';
+import defaultFormValues, {
+  IDMappingFields,
+  IDMappingFormValue,
+  IDMappingFormValues,
+} from '../config/idMappingFormData';
+import { LocationToPath, Location } from '../../../app/config/urls';
+
+import { JobTypes } from '../../types/toolsJobTypes';
+import { addMessage } from '../../../messages/state/messagesActions';
+import {
+  MessageFormat,
+  MessageLevel,
+} from '../../../messages/types/messagesTypes';
+import {
+  IDMappingFormConfig,
+  IDMappingRule,
+  IDMappingField,
+} from '../types/idMappingFormConfig';
+import { FormParameters } from '../types/idMappingFormParameters';
+import { SelectedTaxon } from '../../types/toolsFormData';
+
+import '../../../shared/styles/sticky.scss';
+import '../../styles/ToolsForm.scss';
+import idMappingFormStyle from './style/id-mapping-form.module.scss';
+
+export type TreeDataNode = {
+  label: string;
+  id: string;
+};
+
+export type TreeData = Array<TreeDataNode & { items?: Array<TreeDataNode> }>;
+
+export type DbNameToDbInfo = {
+  [dbName: string]: IDMappingField;
+};
+
+export type RuleIdToRuleInfo = {
+  [ruleID: number]: IDMappingRule;
+};
+
+interface CustomLocationState {
+  parameters?: Partial<FormParameters>;
+}
+
+const IDMappingForm = () => {
+  // refs
+  const fileInputRef = useRef<HTMLInputElement>(null);
+
+  // hooks
+  const dispatch = useDispatch();
+  const history = useHistory();
+  const reducedMotion = useReducedMotion();
+
+  // state
+  const initialFormValues = useMemo(() => {
```

Comment:
We repeat this in a few places in the code base where the only thing that changes is the types. What do you think about making this a custom hook?

Reply:
Sounds good, up to you if you want to have this in this PR or do a different one. Now that we have the 4 forms more or less ready it's a good time to start thinking about what can be refactored and put in common

Reply:
Will make a separate PR for this